config_parser.cc

```diff
@@ -198,18 +200,28 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
       config_stack.top()->statements_.back().get()->child_block_.reset(
           new_config);
       config_stack.push(new_config);
+      block_lvl++;
     } else if (token_type == TOKEN_TYPE_END_BLOCK) {
-      if (last_token_type != TOKEN_TYPE_STATEMENT_END) {
+      // in addition to checking statement_end, also need to check block_end
+      if (last_token_type != TOKEN_TYPE_STATEMENT_END && last_token_type != TOKEN_TYPE_END_BLOCK) {
         // Error.
         break;
       }
       config_stack.pop();
+      if(block_lvl-- < 0) {
```

Comment:
This is a nice early-exit condition. You subtract ` -1` for every closing brace, so if the counter is ever negative, then we've closed more blocks than we've opened. And it's too late to open new blocks—that wouldn't resolve the previous mismatching. Cool work!